File: go/entity/med_supply.go
package entity

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"
)

type MedSupply struct {
	DoseForm                 string
	GenericName              string
	Unit                     string
	YjCode                   string
	YjBase                   string
	Maker                    string
	BrandName                string
	SalesCategory            string
	ShipmentStatus           string
	SupplyStatus             string
	ExpectLiftingStatus      string
	ExpectLiftingDescription string
	Reason                   string
	UpdatedAt                time.Time
}

func MedSupplyFromRow(row []string) MedSupply {
	var arr [17]string
	copy(arr[:], row)

	t := GetTime(arr[13])
	ship := ConvertShipStatus(arr[15])
	sup := ConvertSupplyStatus(arr[11])
	ls := ConvertLiftingStatus(arr[13])
	r := ConvertReason(arr[12])

	return MedSupply{
		DoseForm:                 row[0],
		GenericName:              row[1],
		Unit:                     row[2],
		YjCode:                   row[3],
		YjBase:                   GetYjBase(row[3]),
		Maker:                    row[5],
		BrandName:                arr[4],
		SalesCategory:            arr[6],
		ShipmentStatus:           ship,
		SupplyStatus:             sup,
		ExpectLiftingStatus:      ls,
		ExpectLiftingDescription: arr[11],
		Reason:                   r,
		UpdatedAt:                t,
	}
}

func GetTime(timeString string) time.Time {
	// re, _ := regexp.Compile(`\b(?:Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Sunday),\s(January|February|March|April|May|June|July|August|September|October|November|December)\s(\d{2}),\s(\d{4})\b`)
	// matches := re.FindStringSubmatch(timeString)

	// monthMap := map[string]int{
	// 	"January":   1,
	// 	"February":  2,
	// 	"March":     3,
	// 	"April":     4,
	// 	"May":       5,
	// 	"June":      6,
	// 	"July":      7,
	// 	"August":    8,
	// 	"September": 9,
	// 	"October":   10,
	// 	"November":  11,
	// 	"December":  12,
	// }


	// if matches != nil && len(matches) > 2 {
	// 	year, _ := strconv.Atoi(matches[3])
	// 	month := monthMap[matches[1]] // 月の名前を数字に変換
	// 	day, _ := strconv.Atoi(matches[2])

	// 	return time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
	// }
	// return time.Time{}

	re, _ := regexp.Compile(`(\d{4})年(\d{1,2})月(\d{1,2})日`)
	r := re.FindAllStringSubmatch(timeString, -1)
	if len(r) == 0 {
		return time.Time{}
	}

	d := r[0]

	y, _ := strconv.Atoi(d[1])
	m, _ := strconv.Atoi(d[2])
	da, _ := strconv.Atoi(d[3])

	return time.Date(y, time.Month(m), da, 0, 0, 0, 0, time.UTC)

}

func GetYjBase(yjCode string) string {
	if yjCode == "" {
		return ""
	}

	re, _ := regexp.Compile(`^(\d+)`)
	return re.FindAllStringSubmatch(yjCode, -1)[0][0]
}

func ConvertShipStatus(s string) string {
	if s == "" {
		return ""
	}

	if strings.Contains(s, "Aプラス") {
		return "A+"
	}

	if strings.Contains(s, "A") {
		return "A"
	}

	if strings.Contains(s, "B") {
		return "B"
	}

	if strings.Contains(s, "C") {
		return "C"
	}

	if strings.Contains(s, "D") {
		return "D"
	}

	return ""
}

func ConvertSupplyStatus(s string) string {
	if s == "" {
		return ""
	}

	if strings.Contains(s, "①") {
		return "normal"
	}

	if strings.Contains(s, "②") {
		return "limitedSelf"
	}

	if strings.Contains(s, "③") {
		return "limitedOpponent"
	}

	if strings.Contains(s, "④") {
		return "limitedOther"
	}

	if strings.Contains(s, "⑤") {
		return "stop"
	}

	return ""
}

func ConvertLiftingStatus(s string) string {
	if s == "" {
		return ""
	}

	if strings.Contains(s, "ア") {
		return "promising"
	}

	if strings.Contains(s, "イ") {
		return "promisingNot"
	}

	// ウ、エ、未記入すべてpendingにしておく
	return "pending"
}

func ConvertReason(s string) string {
	if s == "" {
		return ""
	}

	if strings.Contains(s, "１") {
		return "increaseDemand"
	}

	if strings.Contains(s, "２") {
		return "materialShortage"
	}

	if strings.Contains(s, "３") {
		return "manufacturingTrouble"
	}

	if strings.Contains(s, "４") {
		return "qualityTrouble"
	}

	if strings.Contains(s, "５") {
		return "administrativeAction"
	}

	if strings.Contains(s, "６") {
		return "remove"
	}

	// 7, nil は未記入としておく
	return ""
}

func (m MedSupply) ToString() string {
	return fmt.Sprintf("DoseForm: %v, GenericName: %v, Unit: %v, YjCode: %v, YjBase: %v, Maker: %v, BrandName: %v, SalesCategory: %v, ShipmentStatus: %v, SupplyStatus: %v, ExpectLiftingStatus: %v, ExpectLiftingDescription: %v, Reason: %v, UpdatedAt: %v",
		m.DoseForm, m.GenericName, m.Unit, m.YjCode, m.YjBase, m.Maker, m.BrandName, m.SalesCategory, m.ShipmentStatus, m.SupplyStatus, m.ExpectLiftingStatus, m.ExpectLiftingDescription, m.Reason, m.UpdatedAt)
}

File: flutter_med_supply/lib/entity/med_supply.dart
import "package:flutter/material.dart";
import "package:flutter_med_supply/const/const.dart";
import "package:intl/intl.dart";

class MedSupply {
  final int id;
  final String doseForm;
  final String genericName;
  final String brandName;
  final String unit;
  final String yjCode;
  final int yjBase;
  final String maker;
  final String salesCategory; // 先発 or 後発
  final String shipmentStatus;
  final String supplyStatus;
  final String expectLiftingStatus;
  final String expectLiftingDescription;
  final String reason;
  final DateTime updatedAt;
  final String url;
  final String faviconUrl;

  MedSupply({
    required this.id,
    required this.doseForm,
    required this.genericName,
    required this.brandName,
    required this.unit,
    required this.yjCode,
    required this.yjBase,
    required this.maker,
    required this.salesCategory,
    required this.shipmentStatus,
    required this.supplyStatus,
    required this.expectLiftingStatus,
    required this.expectLiftingDescription,
    required this.reason,
    required this.updatedAt,
    required this.url,
    required this.faviconUrl,
  });

  @override
  String toString() {
    return 'MedSupply(id: $id, doseForm: $doseForm, genericName: $genericName, brandName: $brandName, unit: $unit, yjCode: $yjCode, yjBase: $yjBase, maker: $maker, salesCategory: $salesCategory, shipmentStatus: $shipmentStatus, supplyStatus: $supplyStatus, expectLiftingStatus: $expectLiftingStatus, expectLiftingDescription: $expectLiftingDescription, reason: $reason, updatedAt: $updatedAt, url: $url, faviconUrl: $faviconUrl)';
  }

  static final dateFormatter = DateFormat("yyyy-MM-dd hh:mm:ss");

  factory MedSupply.fromJson(Map<String, dynamic> json) {
    final updatedAt = dateFormatter.parse(json['updated_at'] as String);

    return MedSupply(
      id: json['id'] as int,
      doseForm: json['dose_form'] as String,
      genericName: json['generic_name'] as String,
      brandName: json['brand_name'] as String,
      unit: json['unit'] as String,
      yjCode: json['yj_code'] as String,
      yjBase: json['yj_base'] as int,
      maker: json['maker'] as String,
      salesCategory: json['sales_category'] as String,
      shipmentStatus: json['shipment_status'] as String,
      supplyStatus: json['supply_status'] as String,
      expectLiftingStatus: json['expect_lifting_status'] as String,
      expectLiftingDescription: json['expect_lifting_description'] as String,
      reason: json['reason'] as String,
      updatedAt: updatedAt,
      url: json['url'] as String,
      faviconUrl: json['favicon_url'] as String,
    );
  }

  bool isGeneric() {
    return salesCategory == '後発品';
  }

  Color shipmentStatusColor() {
    switch (shipmentStatus) {
      case 'A+':
        return kBlue;
      case 'A':
        return kBlue;
      case 'B':
        return kGreen;
      case 'C':
        return kYellow;
      case 'D':
        return kYellow;
    }
    return kYellow;
  }

  String shipmentStatusString() {
    switch (shipmentStatus) {
      case 'A+':
        return '出荷量増加';
      case 'A':
        return '通常出荷';
      case 'B':
        return '出荷量減少';
      case 'C':
        return '出荷停止';
      case 'D':
        return '販売中止';
    }
    return '出荷情報なし';
  }

  String supplyStatusString() {
    switch (supplyStatus) {
      case "normal":
        return "通常供給";
      case "limitedSelf":
        return "限定供給（自社の事情）";
      case "limitedOpponent":
        return "限定供給（他社品不足のため）";
      case "limitedOther":
        return "限定供給（季節/災害/需要過多による）";
      case "stop":
        return "供給停止";
      default:
        return "供給情報なし";
    }
  }
}
